### PR TITLE
Implemented bulk deletion of expenses excluding corporate card expenses

### PR DESCRIPTION
**Pull request by Dimple16**
Bulk deletion of expenses:

When there is one deletable expense:

Screenshot 2022-03-14 at 10 11 55 PM

When there are multiple deletable expenses:

Screenshot 2022-03-14 at 10 12 07 PM

When there are both deletable expense and corporate card expense in the selection:

Screenshot 2022-03-14 at 10 12 27 PM

When there is only corporate card expenses selected:

Screenshot 2022-03-14 at 10 12 39 PM

### CODE REVIEW

**Comment by Dimple16** [comment]
Few things from the screenshots:
- The primary cta should be `Exclude and delete` in case of trying to delete a combination of expenses
- When the user selects only ccc expenses to delete, the cta - `Delete` is enabled, as these cannot be deleted, we should ideally be disabling the delete button or not show it, verify by design and web app

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 1]
```diff
@@ -1706,23 +1711,78 @@ export class MyExpensesPage implements OnInit {
...
+  }
+
+  getDeletableTxns(expenses) {
+    return expenses.filter((expense) => expense && expense.tx_user_can_delete);
+  }
+
+  getDeleteDialogBody(expensesTobeDeleted, cccExpenses, expenseDeletionMessage, cccExpensesMessage) {
+    let dialogBody: string;
+
+    if (expensesTobeDeleted.length > 0 && cccExpenses > 0) {
```
[the remark this answers is not in the document]
Comment (saitejasFyle) [reply src/app/fyle/my-expenses/my-expenses.page.ts:1725]: Done

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 2]
```diff
@@ -1706,23 +1711,78 @@ export class MyExpensesPage implements OnInit {
...
+    }`;
+  }
+
   async deleteSelectedExpenses() {
     let offlineExpenses: Expense[];
+    let expensesTobeDeleted = this.getDeletableTxns(this.selectedElements);
+    if (this.isUnifyCCCExpensesSettings) {
+      expensesTobeDeleted = this.excludeCCCExpenses(this.selectedElements);
+    }
+    const cccExpenses = this.selectedElements.length - expensesTobeDeleted.length;
```
[the remark this answers is not in the document]
Comment (saitejasFyle) [reply src/app/fyle/my-expenses/my-expenses.page.ts:1766]: Done

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 3]
```diff
@@ -1706,23 +1711,78 @@ export class MyExpensesPage implements OnInit {
...
       componentProps: {
         header: 'Delete Expense',
-        body: `Are you sure you want to delete ${
-          this.selectedElements.length === 1 ? '1 expense?' : this.selectedElements.length + ' expenses?'
-        }`,
+        body: this.getDeleteDialogBody(expensesTobeDeleted, cccExpenses, expenseDeletionMessage, cccExpensesMessage),
+        disableDeleteButton: expensesTobeDeleted.length > 0 ? false : true,
         deleteMethod: () => {
-          offlineExpenses = this.selectedElements.filter((exp) => !exp.tx_id);
+          offlineExpenses = expensesTobeDeleted.filter((exp) => !exp.tx_id);
```
[the remark this answers is not in the document]
Comment (saitejasFyle) [reply src/app/fyle/my-expenses/my-expenses.page.ts:1781]: Done

File: src/global.scss [hunk 4]
```diff
@@ -927,12 +927,23 @@ ref - https://stackoverflow.com/a/31162426
...
   --width: 296px;
-  --max-height: 266px;
+  --max-height: 286px;
   .popover-content {
     border-radius: 16px;
   }
 }
 
+.fy-delete-dialog--body ul {
+  padding: 0px 0px 0px 14px;
```
[the remark this answers is not in the document]
Comment (saitejasFyle) [reply src/global.scss:937]: Done

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 5]
```diff
@@ -1706,23 +1711,78 @@ export class MyExpensesPage implements OnInit {
...
+    if (expensesTobeDeleted.length > 0 && cccExpenses > 0) {
+      dialogBody = `<ul>
+        <li>${cccExpensesMessage}</li>
+        <li>This cannot be undone</li>
+        </ul>
+        <p class="confirmation-message">Do you wish to continue?</p>`;
+    } else if (expensesTobeDeleted.length > 0 && cccExpenses === 0) {
+      dialogBody = `<ul>
+      <li>${expenseDeletionMessage}</li>
+      <li>This cannot be undone</li>
```
[the remark this answers is not in the document]
Comment (saitejasFyle) [reply src/app/fyle/my-expenses/my-expenses.page.ts:1734]: Done

File: src/global.scss [hunk 6]
```diff
@@ -927,12 +927,23 @@ ref - https://stackoverflow.com/a/31162426
...
+  margin: 0;
+  text-align: left;
```
[the remark this answers is not in the document]
Comment (saitejasFyle) [reply src/global.scss:939]: Done

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 7]
```diff
@@ -1706,23 +1711,78 @@ export class MyExpensesPage implements OnInit {
...
 
+  excludeCCCExpenses(expenses) {
+    return expenses.filter((expense) => expense && !expense.tx_corporate_credit_card_expense_group_id);
+  }
+
+  getDeletableTxns(expenses) {
+    return expenses.filter((expense) => expense && expense.tx_user_can_delete);
+  }
+
+  getDeleteDialogBody(expensesTobeDeleted, cccExpenses, expenseDeletionMessage, cccExpensesMessage) {
```
[the remark this answers is not in the document]
Comment (saitejasFyle) [reply src/app/fyle/my-expenses/my-expenses.page.ts:1722]: Done

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 8]
```diff
@@ -1706,23 +1711,78 @@ export class MyExpensesPage implements OnInit {
...
+    } else if (expensesTobeDeleted.length === 0 && cccExpenses > 0) {
+      dialogBody = `<ul>
+      <li>${cccExpensesMessage}</li>
+      </ul>`;
+    }
+
+    return dialogBody;
+  }
+
+  getExpenseDeletionMessage(expensesTobeDeleted) {
```
[the remark this answers is not in the document]
Comment (saitejasFyle) [reply src/app/fyle/my-expenses/my-expenses.page.ts:1746]: Done

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 9]
```diff
@@ -1706,24 +1725,85 @@ export class MyExpensesPage implements OnInit {
...
+  getDeleteDialogBody(
+    expensesToBeDeleted: Expense[],
+    cccExpenses: number,
+    expenseDeletionMessage: string,
+    cccExpensesMessage: string
+  ) {
+    let dialogBody: string;
+
+    if (expensesToBeDeleted?.length > 0 && cccExpenses > 0) {
+      dialogBody = `<ul class="text-left">
```
Comment (Julias0) [inline src/app/fyle/my-expenses/my-expenses.page.ts:1745]: lets move all of this into some service to reduce code volume

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 10]
```diff
@@ -1706,24 +1725,85 @@ export class MyExpensesPage implements OnInit {
     await actionSheet.present();
   }
 
+  excludeCCCExpenses(expenses: Expense[]) {
+    return expenses.filter((expense) => expense && !expense.tx_corporate_credit_card_expense_group_id);
+  }
+
+  getDeletableTxns(expenses: Expense[]) {
```
Comment (Julias0) [inline src/app/fyle/my-expenses/my-expenses.page.ts:1732]: this and the above can be moved into a service

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 11]
```diff
@@ -1706,24 +1725,85 @@ export class MyExpensesPage implements OnInit {
...
+  }
+
+  getExpenseDeletionMessage(expensesToBeDeleted: Expense[]) {
+    return `You are about to permanently delete ${
+      expensesToBeDeleted?.length === 1 ? '1 selected expense.' : expensesToBeDeleted?.length + ' selected expenses.'
+    }`;
+  }
+
+  getCCCExpenseMessage(expensesToBeDeleted: Expense[], cccExpenses: number) {
+    return `There ${cccExpenses > 1 ? ' are ' : ' is '} ${cccExpenses} corporate card ${
```
Comment (Julias0) [inline src/app/fyle/my-expenses/my-expenses.page.ts:1772]: same as above

File: src/app/fyle/my-expenses/my-expenses.page.ts (hunk 9 above)
Comment (saitejasFyle) [reply src/app/fyle/my-expenses/my-expenses.page.ts]: Done

File: src/app/fyle/my-expenses/my-expenses.page.ts (hunk 10 above)
Comment (saitejasFyle) [reply src/app/fyle/my-expenses/my-expenses.page.ts]: Done

File: src/app/fyle/my-expenses/my-expenses.page.ts (hunk 11 above)
Comment (saitejasFyle) [reply src/app/fyle/my-expenses/my-expenses.page.ts]: Done

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 12]
```diff
@@ -1335,6 +1339,18 @@ export class MyExpensesPage implements OnInit {
       this.selectedElements.push(expense);
     }
     this.isReportableExpensesSelected = this.transactionService.getReportableExpenses(this.selectedElements).length > 0;
+
+    if (this.selectedElements?.length > 0) {
+      this.expensesToBeDeleted = this.transactionService.getDeletableTxns(this.selectedElements);
+
+      this.isUnifyCCCExpensesSettings = true;
```
[the remark this answers is not in the document]
Comment (saitejasFyle) [reply src/app/fyle/my-expenses/my-expenses.page.ts]: It's mistake... I've removed it

File: src/app/fyle/my-expenses/my-expenses.page.ts [hunk 13]
```diff
@@ -1335,6 +1339,18 @@ export class MyExpensesPage implements OnInit {
...
+
+      if (this.isUnifyCCCExpensesSettings) {
```
[the remark this answers is not in the document]
Comment (saitejasFyle) [reply src/app/fyle/my-expenses/my-expenses.page.ts:1346]: Same as above

**Pull request merged**